### PR TITLE
🔄 Add rotate example and Axes class

### LINKED ISSUE #1363: vtkAxes not supported by pyvista
Hi Guys 
I want to add axes at given point and then rotate/translate about a given axes 
i tried making a vtk axes but pyvista doesnt seem to support one
is there any solution to this problem ?
axes

### PR DESCRIPTION
**Pull request by tkoyama010**
### Overview

Rotations of a cow about her axes. In this model, the x axis is from the left
to right; the y axis is from bottom to top; and the z axis emerges from the
image. The camera location is the same in all four images.

 
See #1363

TODO:
- [x] Add tests.
- [x] Check that this issue resolves #1363

#### Define camera and axes

Define camera and axes. Setting axes origin to (1.0, 1.0, 1.0).

```python
mesh = examples.download_cow()

camera = pv.Camera()
camera.position = (30.0, 30.0, 30.0)
camera.focal_point = (5.0, 5.0, 5.0)

axes = pv.Axes(show_actor=True)
axes.origin = (3.0, 3.0, 3.0)
```

#### Original Mesh

Plot original mesh. Set axes actor to Plotter.

```python
p = pv.Plotter()

p.add_text("Mesh", font_size=24)
p.add_actor(axes.actor)
p.camera = camera
p.add_mesh(mesh)

p.show()
```
mesh

#### Rotate about x axis

Plot meshs rotated about x axis. It is plotted every 60 degrees.
Set axes actor to Plotter and set axes origin to the point to rotate.

```python
p = pv.Plotter()

p.add_text("X axis rotate", font_size=24)
p.add_actor(axes.actor)
p.camera = camera

for i in range(6):
    rot = mesh.copy()
    rot.rotate_x(60*i, point=axes.origin)
    p.add_mesh(rot)

p.show()
```
rotate_x

#### Rotate about y axis

Plot meshs rotated about y axis. It is plotted every 60 degrees.
Set axes actor to Plotter and set axes origin to the point to rotate.

```python
p = pv.Plotter()

p.add_text("Y axis rotate", font_size=24)
p.camera = camera
p.add_actor(axes.actor)

for i in range(6):
    rot = mesh.copy()
    rot.rotate_y(60*i, point=axes.origin)
    p.add_mesh(rot)

p.show()
```
rotate_y

#### Rotate about z axis

Plot meshs rotated about z axis. It is plotted every 60 degrees.
Set axes actor to Plotter and set axes origin to the point to rotate.

```python
p = pv.Plotter()

p.add_text("Z axis rotate", font_size=24)
p.camera = camera
p.add_actor(axes.actor)

for i in range(6):
    rot = mesh.copy()
    rot.rotate_z(60*i, point=axes.origin)
    p.add_mesh(rot)

p.show()
```
rotate_z

#### Rotate about custom vector

Plot meshs rotated about custom vector. It is plotted every 60 degrees.
Set axes actor to Plotter and set axes origin to the point to rotate.

```python
p = pv.Plotter()

p.add_text("Custom vector rotate", font_size=24)
p.camera = camera
p.add_actor(axes.actor)
for i in range(6):
    rot = mesh.copy()
    rot.rotate_vector(vector=(1, 1, 1), angle=60*i, point=axes.origin)
    p.add_mesh(rot)

p.show()
```
rotate_custom

### CODE REVIEW

File: pyvista/plotting/axes.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    @origin.setter
+    def origin(self, value):
+        """Set the origin of the camera.
+
+        Examples
+        --------
+        >>> import pyvista
+        >>> pl = pyvista.Plotter()
+        >>> pl.axes.origin = (2.0, 1.0, 1.0)
```
Comment (akaszynski) [inline pyvista/plotting/axes.py:57]:
```suggestion
        --------
        >>> import pyvista
        >>> pl = pyvista.Plotter()
        >>> pl.axes.origin
        (0.0, 0.0, 0.0)

        Set the origin of the camera.

        >>> pl.axes.origin = (2.0, 1.0, 1.0)
        >>> pl.axes.origin
        (2.0, 1.0, 1.0)

        """
        return self.GetOrigin()

    @origin.setter
    def origin(self, value):
    """Set the origin of the camera.
```

File: pyvista/plotting/axes.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import pyvista
+from pyvista import _vtk
+
+
+class Axes(_vtk.vtkAxes):
+    """PyVista wrapper for the VTK Axes class.
+
+    Examples
+    --------
+    Create a axes at the pyvista module level
```
Comment (akaszynski) [inline pyvista/plotting/axes.py:14]:
```suggestion
    Create an instance of axes at the pyvista module level.
```

File: pyvista/core/dataset.py [hunk 3]
```diff
@@ -466,40 +466,101 @@ def points_to_double(self):
...
+    def rotate_z(self, angle: float, point=None, transform_all_input_vectors=False):
         """Rotate mesh about the z-axis.
 
         Parameters
         ----------
         angle : float
             Angle in degrees to rotate about the z-axis.
 
+        point : float, optional
+            Point to ratate about.
```
Comment (akaszynski) [inline pyvista/core/dataset.py:524]:
```suggestion
            Point to ratate about.
```
```suggestion
            Point to rotate about.
```

File: pyvista/core/dataset.py [hunk 4]
```diff
@@ -466,40 +466,101 @@ def points_to_double(self):
...
-    def rotate_z(self, angle: float, transform_all_input_vectors=False):
+    def rotate_z(self, angle: float, point=None, transform_all_input_vectors=False):
         """Rotate mesh about the z-axis.
 
         Parameters
         ----------
         angle : float
             Angle in degrees to rotate about the z-axis.
 
+        point : float, optional
```
Comment (akaszynski) [inline pyvista/core/dataset.py:523]:
```suggestion
        point : list, optional
```

File: pyvista/core/dataset.py [hunk 5]
```diff
@@ -466,40 +466,101 @@ def points_to_double(self):
...
+        Parameters
+        ----------
+        vector : tuple
+            Axes to rotate about.
+
+        angle : float
+            Angle in degrees to rotate about the vector.
+
+        point : float, optional
+            Point to ratate about.
```
Comment (akaszynski) [inline pyvista/core/dataset.py:550]:
```suggestion
        point : list, optional
            Point to rotate about.  Defaults to origin ``(0.0, 0.0, 0.0)``.
```

File: pyvista/core/dataset.py [hunk 6]
```diff
@@ -466,40 +466,101 @@ def points_to_double(self):
...
+    def rotate_x(self, angle: float, point=None, transform_all_input_vectors=False):
         """Rotate mesh about the x-axis.
 
         Parameters
         ----------
         angle : float
             Angle in degrees to rotate about the x-axis.
 
+        point : float, optional
+            Point to ratate about.
```
Comment (akaszynski) [inline pyvista/core/dataset.py:478]:
```suggestion
        point : list, optional
            Point to rotate about.  Defaults to origin ``(0.0, 0.0, 0.0)``.
```

File: examples/01-filter/rotate.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
+"""
+Rotations
+~~~~~~~~~
+
+Rotations of a cow about her axes. In this model, the x axis is from the left
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:5]:
```suggestion
Rotations of a mesh about its axes. In this model, the x axis is from the left
```

File: examples/01-filter/rotate.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+"""
+# sphinx_gallery_thumbnail_number = 6
+import pyvista as pv
+from pyvista import examples
+
+###############################################################################
+# Define camera and axes
+# ++++++++++++++++++++++
+#
+# Define camera and axes. Setting axes origin to (1.0, 1.0, 1.0).
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:18]:
```suggestion
# Define camera and axes. Setting axes origin to ``(1.0, 1.0, 1.0)``.
```

**Comment by akaszynski** [comment]
Cool example @tkoyama010!  I've added some additional parameters to the new Axes class and changed the scale of the cow to make the figures look a little better.  Please let me know if you agree with the changes.

File: pyvista/plotting/axes.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        >>> pl.axes.origin
+        (2.0, 1.0, 1.0)
+
+        """
+        return self.GetOrigin()
+
+    @origin.setter
+    def origin(self, value):
+    """Set the origin of the camera.
+        """
```
Comment (akaszynski) [inline pyvista/plotting/axes.py:72]:
```suggestion
    """Set the origin of the camera."""
```

**Review by tkoyama010 — COMMENTED** [review]
@akaszynski  Thanks. Adding `actor_scale` and `line_width` is good idea 😀

File: pyvista/plotting/axes.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        Examples
+        --------
+        >>> import pyvista as pv
+        >>> axes = pv.Axes()
+        >>> axes.show_symmetric()
+        """
+        self.SymmetricOn()
+
+    def hide_symmetric(self):
+        """Show symmetric of axes.
```
Comment (tkoyama010) [inline pyvista/plotting/axes.py:112]:
```suggestion
        """Hide symmetric of axes.
```

**Review by tkoyama010 — COMMENTED** [review]
WFM

File: examples/01-filter/rotate.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+"""
+# sphinx_gallery_thumbnail_number = 6
+import pyvista as pv
+from pyvista import examples
+
+###############################################################################
+# Define camera and axes
+# ++++++++++++++++++++++
+#
+# Define camera and axes. Setting axes origin to ``(1.0, 1.0, 1.0)``.
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:18]:
```suggestion
# Define camera and axes. Setting axes origin to ``(3.0, 3.0, 3.0)``.
```

File: examples/01-filter/rotate.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+camera.focal_point = (5.0, 5.0, 5.0)
+
+axes = pv.Axes(show_actor=True, actor_scale=2.0, line_width=5)
+axes.origin = (3.0, 3.0, 3.0)
+
+###############################################################################
+# Original Mesh
+# +++++++++++++
+#
+# Plot original mesh. Set axes actor to Plotter.
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:34]:
```suggestion
# Plot original mesh. Add axes actor to Plotter.
```

File: examples/01-filter/rotate.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+p.camera = camera
+p.add_mesh(mesh)
+
+p.show()
+
+###############################################################################
+# Rotate about x axis
+# +++++++++++++++++++
+#
+# Plot meshs rotated about x axis. It is plotted every 60 degrees.
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:49]:
```suggestion
# Plot the mesh rotated about x axis in 60 degree increments.
```

File: examples/01-filter/rotate.py [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+# Set axes actor to Plotter and set axes origin to the point to rotate.
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:50]:
```suggestion
# Add axes actor to Plotter and set axes origin to the point to rotate.
```

File: examples/01-filter/rotate.py [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+
+p = pv.Plotter()
+
+p.add_text("X axis rotate", font_size=24)
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:54]:
```suggestion
p.add_text("X-Axis Rotation", font_size=24)
```

File: examples/01-filter/rotate.py [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+    p.add_mesh(rot)
+
+p.show()
+
+###############################################################################
+# Rotate about y axis
+# +++++++++++++++++++
+#
+# Plot meshs rotated about y axis. It is plotted every 60 degrees.
+# Set axes actor to Plotter and set axes origin to the point to rotate.
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:70]:
```suggestion
# Plot the rotated about the y axis every 60 degrees.
# Add the axes actor to the Plotter and set axes origin to the point to rotation.
```

File: examples/01-filter/rotate.py [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+    p.add_mesh(rot)
+
+p.show()
+
+###############################################################################
+# Rotate about z axis
+# +++++++++++++++++++
+#
+# Plot meshs rotated about z axis. It is plotted every 60 degrees.
+# Set axes actor to Plotter and set axes origin to the point to rotate.
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:90]:
```suggestion
# Plot the mesh rotated about the z axis every 60 degrees.
# Add axes actor to the Plotter and set the axes origin to the point of rotation.
```

File: examples/01-filter/rotate.py [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+
+p = pv.Plotter()
+
+p.add_text("Z axis rotate", font_size=24)
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:94]:
```suggestion
p.add_text("Z-Axis Rotation", font_size=24)
```

File: examples/01-filter/rotate.py (hunk 14 above)
Comment (akaszynski) [inline examples/01-filter/rotate.py:50]:
```suggestion
# Plot the mesh rotated about the x axis every 60 degrees.
# Add the axes actor to the Plotter and set the axes origin to the point of rotation.
```

File: examples/01-filter/rotate.py [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+p.add_text("Mesh", font_size=24)
+p.add_actor(axes.actor)
+p.camera = camera
+p.add_mesh(mesh)
+
+p.show()
+
+###############################################################################
+# Rotate about x axis
+# +++++++++++++++++++
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:47]:
```suggestion
# Rotation about the x axis
# +++++++++++++++++++++++++
```

File: examples/01-filter/rotate.py (hunk 15 above)
Comment (akaszynski) [inline examples/01-filter/rotate.py:54]:
```suggestion
p.add_text("X-Axis Rotation", font_size=24)
```

File: examples/01-filter/rotate.py [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+for i in range(6):
+    rot = mesh.copy()
+    rot.rotate_x(60*i, point=axes.origin)
+    p.add_mesh(rot)
+
+p.show()
+
+###############################################################################
+# Rotate about y axis
+# +++++++++++++++++++
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:67]:
```suggestion
# Rotation about the y axis
# +++++++++++++++++++++++++
```

File: examples/01-filter/rotate.py [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+#
+# Plot meshs rotated about y axis. It is plotted every 60 degrees.
+# Set axes actor to Plotter and set axes origin to the point to rotate.
+
+p = pv.Plotter()
+
+p.add_text("Y axis rotate", font_size=24)
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:74]:
```suggestion
p.add_text("Y-Axis Rotation", font_size=24)
```

File: examples/01-filter/rotate.py [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+for i in range(6):
+    rot = mesh.copy()
+    rot.rotate_y(60*i, point=axes.origin)
+    p.add_mesh(rot)
+
+p.show()
+
+###############################################################################
+# Rotate about z axis
+# +++++++++++++++++++
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:87]:
```suggestion
# Rotation about the z axis
# +++++++++++++++++++++++++
```

File: examples/01-filter/rotate.py [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+for i in range(6):
+    rot = mesh.copy()
+    rot.rotate_z(60*i, point=axes.origin)
+    p.add_mesh(rot)
+
+p.show()
+
+###############################################################################
+# Rotate about custom vector
+# ++++++++++++++++++++++++++
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:107]:
```suggestion
# Rotation about a custom vector
# ++++++++++++++++++++++++++++++
```

File: examples/01-filter/rotate.py [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+#
+# Plot meshs rotated about custom vector. It is plotted every 60 degrees.
+# Set axes actor to Plotter and set axes origin to the point to rotate.
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:110]:
```suggestion
# Plot the mesh rotated about a custom vector every 60 degrees.
# Add the axes actor to the Plotter and set axes origin to the point of rotation.
```

File: examples/01-filter/rotate.py [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+
+p = pv.Plotter()
+
+p.add_text("Custom vector rotate", font_size=24)
```
Comment (akaszynski) [inline examples/01-filter/rotate.py:114]:
```suggestion
p.add_text("Custom Vector Rotation", font_size=24)
```

**Review by akaszynski — APPROVED** [review]
Approved with suggestions.  Thanks for your work @tkoyama010!

**Comment by ghazalbangash** [comment]
sorry for bothering again but even though I updated my pyvista and even installed it again using git , I get the following error "AttributeError: module 'pyvista' has no attribute 'Axes' "  and I checked I had the axes.py file in my pyvista folder .Has the same problem with camera

**Comment by akaszynski** [comment]
I had this problem once and it turned out that I hadn't installed pyvista correctly and/or was importing pyvista from my local git directory.

**Comment by ghazalbangash** [comment]
yes I did but since I am using background plotter(pyvistaqt) it would give me the following error on updating so that i could use Axes and camera : AttributeError: plotter has no attribute named _add_observer (checked it both with pyvistaqt 0.4.0 and 0.5.0)
it works fine with pv.plotter

**Comment by akaszynski** [comment]
> yes I did but since I am using background plotter(pyvistaqt) it would give me the following error on updating so that i could use Axes and camera : AttributeError: plotter has no attribute named _add_observer (checked it both with pyvistaqt 0.4.0 and 0.5.0)
it works fine with pv.plotter

Please raise an issue on https://github.com/pyvista/pyvistaqt

**Pull request merged**